example/Index.re

```diff
@@ -1,41 +1,123 @@
+let check_properties req next properties k => {
+  let rec aux properties => {
+    switch properties {
+      | [] => k ();
+      | [p, ...tl] => check_property req next p (fun () => aux tl); 
+    }; 
+  };
+  aux properties;
+};
+
+/* [set_property req property] sets the [property] in the [req] Request 
+   value */
+let set_property req property => { 
+  let reqData = Request.asJsonObject req; 
+  Js_dict.set reqData property (Js_json.boolean Js.true_);
+};
+
+/* sends a common JSON object representing success */
+let send_success_json res => {
+  let json = Js_dict.empty (); 
+  Js_dict.set json "success" (Js_json.boolean Js.true_); 
+  Response.sendJson res (Js_json.object_ json);
+};
+
+
 let app = express ();
 
-App.useOnPath app path::"/" @@ Middleware.from (fun _ _ next => {
-  Js.log "Request received"; 
-     /* This will be printed for every request */
+App.useOnPath app path::"/" @@ Middleware.from (fun req _ next => {
+  set_property req "middleware0";
   next Next.middleware
-     /* call the next middleware in the processing pipeline */
+    /* call the next middleware in the processing pipeline */
 });
 
 App.useN app [|
-  Middleware.from (fun _ _ next => {
-    Js.log "trace #1";
-    next Next.middleware
+  Middleware.from (fun req _ next => {
+    check_property req next "middleware0" (fun () => {
+      set_property req "middleware1";
+      next Next.middleware;
+    }); 
   }),
-  Middleware.from (fun _ _ next => {
-    Js.log "trace #2";
-    next Next.middleware
+
+  Middleware.from (fun req _ next => {
+    check_properties req next ["middleware0", "middleware1"] (fun () => {
+      set_property req "middleware2";
+      next Next.middleware;
+    }); 
   })
 |]; 
 
-App.get app path::"/" @@ Middleware.from (fun _ res _ => { 
-  let json = Js_dict.empty ();
-  Js_dict.set json "hello" (Js_json.string "World"); 
-    /* This only works with dev version of BuckleScript */
-  Response.sendJson res (Js_json.object_ json);
+App.get app path::"/" @@ Middleware.from (fun req res next => { 
+  let prev_middlewares = [ "middleware0", "middleware1", "middleware2"];
+  check_properties req next prev_middlewares (fun () => {
+    send_success_json res;
+  });
 });
 
-App.useOnPathN app path::"/static" @@ [| 
-  Middleware.from (fun _ _ next => {
-    Js.log "trace #3";
-    next Next.middleware
-  }), 
-  {
+App.useOnPath app path::"/static" {
     let options = Static.defaultOptions (); 
     Static.make "static" options |> Static.asMiddleware 
-  }, 
-|]; 
+};
+
+App.postN app path::"/:id/id" [|
+  Middleware.from (fun req res next => {
+    let prev_middlewares = [ "middleware0", "middleware1", "middleware2"];
+    check_properties req next prev_middlewares (fun () => {
+      switch (Js_dict.get (Request.params req) "id") {
+        | Some json => switch (Js_json.decodeString json) {
+          | Some "123" => send_success_json res
+          | _ => next Next.route;
+        };
+        | _ => next Next.route;
```

Comment:
This boilerplate could be nicer, either by abstracting the parsing entirely, or by using a few `Option` functions.